Index: twisted/words/protocols/irc.py
===================================================================
--- twisted/words/protocols/irc.py	(revision 39738)
+++ twisted/words/protocols/irc.py	(working copy)
@@ -2285,7 +2285,7 @@
         if len(data) < 3:
             raise IRCBadMessage, "malformed DCC CHAT request: %r" % (data,)
 
-        (filename, address, port) = data[:3]
+        (protocol, address, port) = data[:3]
 
         address = dccParseAddress(address)
         try:
@@ -2293,7 +2293,7 @@
         except ValueError:
             raise IRCBadMessage, "Indecipherable port %r" % (port,)
 
-        self.dccDoChat(user, channel, address, port, data)
+        self.dccDoChat(user, channel, protocol, address, port, data)
 
     ### The dccDo methods are the slightly higher-level siblings of
     ### common dcc_ methods; the arguments have been parsed for them.
@@ -2320,7 +2320,7 @@
         request made by us.  By default it will do nothing."""
         pass
 
-    def dccDoChat(self, user, channel, address, port, data):
+    def dccDoChat(self, user, channel, protocol, address, port, data):
         pass
         #factory = DccChatFactory(self, queryData=(user, channel, data))
         #reactor.connectTCP(address, port, factory)
